runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/middleware/FlexibleChecksumsRequestMiddleware.kt

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.http.middleware
+
+import aws.smithy.kotlin.runtime.ClientException
+import aws.smithy.kotlin.runtime.hashing.*
+import aws.smithy.kotlin.runtime.http.*
+import aws.smithy.kotlin.runtime.http.operation.*
+import aws.smithy.kotlin.runtime.http.request.header
+import aws.smithy.kotlin.runtime.io.*
+import aws.smithy.kotlin.runtime.util.*
+import kotlinx.coroutines.*
+import kotlin.coroutines.coroutineContext
+
+/**
+ * Mutate a request to enable flexible checksums.
+ *
+ * If the checksum will be sent as a header, calculate the checksum.
+ *
+ * Otherwise, if it will be sent as a trailing header, calculate the checksum as asynchronously as the body is streamed.
+ * In this case, a [LazyAsyncValue] will be added to the execution context which allows the trailing checksum to be sent
+ * after the entire body has been streamed.
+ *
+ * @param checksumAlgorithmName the name of the algorithm used to calculate the checksum
+ */
+@InternalApi
+public class FlexibleChecksumsRequestMiddleware(private val checksumAlgorithmName: String) : ModifyRequestMiddleware {
+    public override suspend fun modifyRequest(req: SdkHttpRequest): SdkHttpRequest {
+        val logger = coroutineContext.getLogger<FlexibleChecksumsRequestMiddleware>()
+
+        val headerName = "x-amz-checksum-${checksumAlgorithmName.lowercase()}"
+        logger.debug { "Resolved header name: $headerName" }
+
+        // remove all checksum headers except for $headerName
+        // this handles the case where a user inputs a precalculated checksum, but it doesn't match the input checksum algorithm
+        req.subject.headers.removeAllChecksumHeadersExcept(headerName)
+
+        val checksumAlgorithm = checksumAlgorithmName.toHashFunction() ?: throw ClientException("Could not parse checksum algorithm $checksumAlgorithmName")
+
+        if (!checksumAlgorithm.isSupported) {
+            throw ClientException("Checksum algorithm $checksumAlgorithmName is not supported for flexible checksums")
+        }
+
+        if (req.subject.body.isEligibleForAwsChunkedStreaming) {
+            req.subject.header("x-amz-trailer", headerName)
+
+            val deferredChecksum = CompletableDeferred<String>()
```

Comment:
Agreed we should tie this to _probably_ the execution context (coroutine context) [job](https://github.com/awslabs/smithy-kotlin/blob/209e835ad85004b72b977a439ef591f570378109/runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/client/ExecutionContext.kt#L21). e.g. `req.context.coroutineContext.job`. It will get cancelled when the request is complete.